src/lib.rs

```diff
@@ -323,8 +326,44 @@ impl fmt::Display for GlobError {
     }
 }
 
-fn is_dir(p: &Path) -> bool {
-    fs::metadata(p).map(|m| m.is_dir()).unwrap_or(false)
+#[derive(Debug)]
+struct PathWrapper {
+    path: PathBuf,
+    is_directory: bool,
+}
+
+impl PathWrapper {
+    fn from_dir_entry(path: PathBuf, e: DirEntry) -> Self {
+        let is_directory = e
+            .file_type()
+            .ok()
+            .map(|file_type| file_type.is_dir())
+            .or_else(|| fs::metadata(&path).map(|m| m.is_dir()).ok())
```

Comment:
I want to complain that this is following symlinks but apparently that's a preexisting issue and there's a separate PR for that. So it's ok for now.

But I that means `is_dir` will be false if it's a symlink. Since the current behavior is to follow symlinks we still have to do `fs::metadata` when `file_type.is_symlink() == true` to preserve the current behavior.

Reply:
It seems there are no tests covering that case?

Reply:
I also wondered how does it handle symlinks.. and then I found out that it doesn't :) Good catch, fixed.

Reply:
Also added a test for it.